### PR TITLE
scaling: work around an upstream bug on wayland

### PR DESCRIPTION
**Pull request by capfredf**
`gdk_screen_get_monitor_geometry` is supposed to return the logic size of a
display. However, due to the upstream bug [1], we have to manually recalculate
the size.

[1] https://gitlab.gnome.org/GNOME/gtk/-/issues/2599

### CODE REVIEW

**Comment by capfredf** [comment]
I ran into this problem when I was running slideshow on a 4k display with a scale@2x on wayland. 
The resulting slides were stretched:

**Comment by mflatt** [comment]
I'm not sure I follow, but it looks like this will produce the right answer with GTK3 >= 3.24.9 and the wrong answer for earlier versions (and perhaps a future repaired version). Is that true, or am I misreading?

**Comment by capfredf** [comment]
Yes, that is correct. If the upstream issue is fixed, the patched function will produce a wrong answer.  As to earlier versions of gtk, we can only apply this patch only with gtk3 >= 3.24.9, but I don't know what is an idiomatic way to do that in `racket/gui`.

**Comment by capfredf** [comment]
I tried `gdk_monitor_get_geometry` when I was working this patch. It produced the same incorrect answer.

**Comment by capfredf** [comment]
`gtk does provide constants GTK_MAJOR_VERSION, GTK_MICRO_VERSION and GTK_MINOR_VERSION`
I took another look. Gtk also provides three functions to get there version numbers.

**Comment by mflatt** [comment]
Just to make sure, does `gdk_monitor_get_workarea` have the same scaling issue?

**Comment by capfredf** [comment]
It produces the same result

```
(define-gdk gdk_monitor_get_workarea (_fun _GdkMonitor _GdkRectangle-pointer -> _void))
(define-gdk gdk_display_get_default (_fun -> _GdkDisplay))
(define-gdk gdk_display_get_monitor (_fun _GdkDisplay _int -> _GdkMonitor))

(let* ([d (gdk_display_get_default)]
       [m (gdk_display_get_monitor d 0)]
       [r (make-GdkRectangle 0 0 0 0)])
  (define scale (gdk_screen_get_monitor_scale_factor
                 (gdk_screen_get_default)
                 0))
  (gdk_monitor_get_workarea m r)
  (eprintf "scale ~a x ~a y ~a width ~a height ~a~n"
           scale
           (GdkRectangle-x r)
           (GdkRectangle-y r)
           (GdkRectangle-width r)
           (GdkRectangle-height r)))
```
output:
```
scale 2 x 3840 y 0 width 3840 height 2160
```

File: gui-lib/mred/private/wx/gtk/frame.rkt [hunk 1]
```diff
@@ -595,7 +599,20 @@
     (unless (num . < . (gdk_screen_get_n_monitors s))
       (fail))
     (gdk_screen_get_monitor_geometry s num r)
-    r))
+    (cond
+      ;; work around the upstream issue https://gitlab.gnome.org/GNOME/gtk/-/issues/2599
+      [(and wayland?
+            (= (gtk_get_major_version) 3)
+            (>= (gtk_get_minor_version) 24)
+            (>= (gtk_get_micro_version) 9))
```
Comment (mflatt) [inline gui-lib/mred/private/wx/gtk/frame.rkt:607]: I think this `and` probably doesn't do what you intend for a minor version that's strictly more than 24.

**Comment by mflatt** [comment]
I never found a better strategy than checking the version, so I'm ok with this approach.

Besides the correction to the version comparison, it appears that these function are not available from Gtk 2. A combination of using `#:fail` in `define-gtk` and using `gtk3?` in the conditional is probably the right idea.

**Pull request merged**